thunder/executors/nvfuserex_impl.py

```diff
@@ -2732,6 +2738,272 @@ def embedding(
+) -> Any:
+    return TensorProxy(like=a)
+
+
+def cross_entropy_bwd(
+    g: TensorLike,
+    a: TensorLike,
+    /,
+    *,
+    target: TensorLike,
+    a_max: TensorLike,
+    max_log_sum_exp: TensorLike,
+    valid_indices: TensorLike,
+    ignore_index: int = -100,
+    label_smoothing: float = 0.0,
+    fd: FusionDefinition,
+    lc_to_nv_map: dict,
+) -> Any:
+    nv_a = getnv(a, fd, lc_to_nv_map)
+    nv_target = getnv(target, fd, lc_to_nv_map)
+    nv_ignore_index = getnv(ignore_index, fd, lc_to_nv_map)
+    nv_g = getnv(g, fd, lc_to_nv_map)
+    nv_valid_indices = getnv(valid_indices, fd, lc_to_nv_map)
+    nv_a_max = getnv(a_max, fd, lc_to_nv_map)
+    nv_max_log_sum_exp = getnv(max_log_sum_exp, fd, lc_to_nv_map)
+
+    zero = fd.define_scalar(0, dtype=DataType.Int)
+    one = fd.define_scalar(1, dtype=DataType.Int)
+
+    # scatter the gradients (negative) - this is backward of nll los
+    iotas = fd.ops.iota(nv_a.shape()[-1], zero, one, dtype=DataType.Int)
+    iotas_bcast = fd.ops.broadcast_in_dim(iotas, shape=nv_a.shape(), broadcast_dims=[nv_a.ndim - 1])
+    neg_gradients = fd.ops.neg(nv_g)
+    neg_gradients_mod = fd.ops.div(neg_gradients, nv_valid_indices)
+
+    target_bcast = fd.ops.broadcast_in_dim(nv_target, shape=nv_a.shape(), broadcast_dims=[nv_target.ndim - 1])
+    mask = fd.ops.eq(iotas_bcast, target_bcast)
+    neg_gradients_mod_bcast = fd.ops.broadcast_in_dim(neg_gradients_mod, shape=nv_a.shape(), broadcast_dims=[])
+    scattered_vals = fd.ops.where(mask, neg_gradients_mod_bcast, zero)
+
+    ne = fd.ops.ne(target_bcast, nv_ignore_index)
+    new_target_bcast = fd.ops.where(ne, neg_gradients_mod_bcast, zero)
+
+    # build the softmax
+    nv_a_max_bcast = fd.ops.broadcast_in_dim(nv_a_max, shape=nv_a.shape(), broadcast_dims=[nv_a_max.ndim - 1])
+    input_minus_max = fd.ops.sub(nv_a, nv_a_max_bcast)
+    nv_max_log_sum_exp_bcast = fd.ops.broadcast_in_dim(
+        nv_max_log_sum_exp, shape=nv_a.shape(), broadcast_dims=[nv_max_log_sum_exp.ndim - 1]
+    )
+    log_softmax = fd.ops.sub(input_minus_max, nv_max_log_sum_exp_bcast)
+    recomputed_softmax = fd.ops.exp(log_softmax)
+
+    # this should be gradient - softmax * gradient_sum
+    softmax_mul_grad_sum = fd.ops.mul(recomputed_softmax, new_target_bcast)
+    difference = fd.ops.sub(scattered_vals, softmax_mul_grad_sum)
+
+    return difference
+
+
+nv_cross_entropy_bwd = ex.register_operator(
+    "nv_cross_entropy_bwd",
+    meta=cross_entropy_bwd_meta,
+    fn=cross_entropy_bwd,
+    tags=[prims.OpTags.RANDOM_OP],
+)
+
+register_supported(nv_cross_entropy_bwd.id, cross_entropy_bwd, None)
+
+
+def cross_entropy_transform(
+    a: TensorLike,
+    /,
+    target: TensorLike,
+    weight: None | TensorLike = None,
+    size_average: None | Any = None,
+    ignore_index: int = -100,
+    reduce: None | Any = None,
+    reduction: str = "mean",
+    label_smoothing: float = 0.0,
+) -> Any:
+    result, _, _, _ = nv_cross_entropy_fwd(
```

Comment:
Just for my own curiosity (trying my luck with @IvanYashchuk @mruberry) 

When is `execution_transform` used? I assume `grad_transform` would kick-in before that.
But if grad transform already decomposed it to `nv_cross_entropy_fwd`, we don't need an `execution_transform`?

I guess inference would be a case where we wouldn't run grad transform, so it's necessary to have this as well. But I wouldn't expect inference workflow running a loss function.

Reply:
The grad transform does happen first and it does decompose to custom operators, like `nv_cross_entropy_fwd`, but those operators aren't assigned to any executors. Then during the execution transform nvFuser will be the only executor that can execute `nv_cross_entropy_fwd` and will be assigned it. 

In general a grad transform could create operations grabbed by a mix of executors, and even an executor's custom grad could create operations that should be executed by a mix of executors.

Reply:
In the no_grad context, the grad transform would not be used, and the execution transform would be used.